### PR TITLE
Pytes battery: Add support for native CAN protocol

### PR DESCRIPTION
**Pull request by ranma**
The recently added PytesCanReceiver.cpp implements the Victron CAN protocol.

This change additionally adds support for the native Pytes CAN protocol messages.

Features only supported in Pytes protocol:
- High-resolution state of charge / full and remaining mAh
- Charge cycle counter
- Balancing state

Features only supported in Victron protocol:
- FW version
- Serial number

Note that the only known way to select the native Pytes protocol is via the serial console (Cisco-compatible cables work):

```
login config
setprt PYTES
logout
```

to return to Victron protocol use:
```
login config
setprt VICTRON
logout
```

The protocol switch will take effect immediately.

Tested on Pytes E-Box 4850

See https://github.com/helgeerbe/OpenDTU-OnBattery/issues/1188

### CODE REVIEW

**Comment by AndreasBoehm** [comment]
Thanks for your PR, cool that we can get more details about the battery by switching it to a different protocol. 
I will test this once my console cable arrives.

Things that we should adjust:
- Figure out what 'balance' actually represents and add it to the UI
- If we can't figure out state/warning flags we should remove the section from the UI, but it also doesn't feel right for me to switch to a different protocol but then we loose important things like warnings
- Remove any information from the UI that is not available anymore

**Comment by schlimmchen** [comment]
> Remove any information from the UI that is not available anymore

The UI is generic, as far as I remember. Only info that is put into the respective JSON array is displayed in the UI. So I am not sure why you mention this. We have to take care that in the backend, only relevant info is serialized into JSON. The respective BatteryProvider implementation is responsible for that.

**Comment by AndreasBoehm** [comment]
@schlimmchen That is exactly what i meant with my comment. Don't touch the UI code, adjust the BatteryStats for 'Pytes' to not add values into the JSON that are not available.

**Comment by ranma** [comment]
Updated the pull request with support for alarms, warnings and online/offline module count.

File: src/BatteryStats.cpp [hunk 1]
```diff
@@ -129,10 +129,13 @@ void PytesBatteryStats::getLiveViewData(JsonVariant& root) const
     addLiveViewValue(root, "dischargeVoltageLimitation", _dischargeVoltageLimit, "V", 1);
     addLiveViewValue(root, "dischargeCurrentLimitation", _dischargeCurrentLimit, "A", 1);
     addLiveViewValue(root, "stateOfHealth", _stateOfHealth, "%", 0);
+    if (_chargeCycles > 0) {
```
Comment (AndreasBoehm) [inline src/BatteryStats.cpp:132]: Let's use `-1` as the default value of `_chargeCycles` to replicate how we defined the 'no-data'sate for `_chargedEnergy` and `_dischargedEnergy`

File: src/BatteryStats.cpp [hunk 2]
```diff
@@ -347,6 +350,8 @@ void PytesBatteryStats::mqttPublish() const
     MqttSettings.publish("battery/settings/dischargeVoltageLimitation", String(_dischargeVoltageLimit));
 
     MqttSettings.publish("battery/stateOfHealth", String(_stateOfHealth));
+    MqttSettings.publish("battery/chargeCycles", String(_chargeCycles));
```
Comment (AndreasBoehm) [inline src/BatteryStats.cpp:353]: Should only be published if we have actually have data

File: src/BatteryStats.cpp [hunk 3]
```diff
@@ -347,6 +350,8 @@ void PytesBatteryStats::mqttPublish() const
...
+    MqttSettings.publish("battery/balance", String(_balance));
```
Comment (AndreasBoehm) [inline src/BatteryStats.cpp:354]: Should only be published if we have actually have data

File: src/BatteryStats.cpp (hunk 1 above)
Comment (AndreasBoehm) [inline src/BatteryStats.cpp:132]: Please also add 'balance' to the LiveView

File: src/PytesCanReceiver.cpp [hunk 4]
```diff
@@ -284,6 +288,163 @@ void PytesCanReceiver::onMessage(twai_message_t rx_message)
...
+            uint32_t alarmBits1 = this->readUnsignedInt32(rx_message.data);
+            uint32_t alarmBits2 = this->readUnsignedInt32(rx_message.data + 4);
+            uint32_t mergedBits = alarmBits1 | alarmBits2;
+
+            bool overVoltage = this->getBit(mergedBits, 0);
+            bool highVoltage = this->getBit(mergedBits, 1);
+            // bool normalVoltage = this->getBit(mergedBits, 2);
+            bool lowVoltage = this->getBit(mergedBits, 3);
+            bool underVoltage = this->getBit(mergedBits, 4);
+            // bool chargeHigh = this->getBit(mergedBits, 5);
```
Comment (AndreasBoehm) [inline src/PytesCanReceiver.cpp:335]: What about the commented bits?

File: src/PytesCanReceiver.cpp [hunk 5]
```diff
@@ -284,6 +288,163 @@ void PytesCanReceiver::onMessage(twai_message_t rx_message)
...
+            bool overCurrentCharge = this->getBit(mergedBits, 19) || this->getBit(mergedBits, 20);
+            bool highCurrentDischarge = this->getBit(mergedBits, 21);
+            bool highCurrentCharge = this->getBit(mergedBits, 22);
+            // bool stateIdle = this->getBit(mergedBits, 25);
+            bool stateCharging = this->getBit(mergedBits, 26);
+            bool stateDischarging = this->getBit(mergedBits, 27);
+
+            _stats->_alarmOverVoltage = overVoltage;
+            _stats->_alarmUnderVoltage = underVoltage;
+            _stats->_alarmOverTemperature = stateDischarging && overTemp;
```
Comment (AndreasBoehm) [inline src/PytesCanReceiver.cpp:353]: what about `stateIdle`?

File: src/PytesCanReceiver.cpp [hunk 6]
```diff
@@ -284,6 +288,163 @@ void PytesCanReceiver::onMessage(twai_message_t rx_message)
...
+                MessageOutput.printf("[Pytes] Warnings: %08x %08x\r\n",
+                        alarmBits1, alarmBits2);
+            }
+            break;
+        }
+
+        case 0x404: { // Pytes protocol: SOC/SOH
+            // soc isn't used here since it is generated with higher
+            // precision in message 0x0409 below.
+            uint8_t soc = static_cast<uint8_t>(this->readUnsignedInt16(rx_message.data));
```
Comment (AndreasBoehm) [inline src/PytesCanReceiver.cpp:381]: i think we can omit parsing the soc if we don't use it

**Review by AndreasBoehm — CHANGES_REQUESTED** [review]

**Comment by ranma** [comment]
With this change and Victron protocol:
screenshot victron

With this change, Pytes protocol and low charge state:
screenshot pytes low charge

With this change, Pytes protocol, fully charged and balancing active:
screenshot pytes fully charged and balancing

**Comment by ranma** [comment]
Added HASS auto-discovery for new fields.

**Comment by AndreasBoehm** [comment]
@ranma could you do me a favor and update the description of the PR to reflect what features/changes we can expect from this PR?
Its a bit hard to follow the comments and the code does not match the description anymore (e.g. support of alarms)

Thank you.

File: src/PytesCanReceiver.cpp [hunk 7]
```diff
@@ -5,6 +5,16 @@
 #include <driver/twai.h>
 #include <ctime>
 
+namespace {
+
+static void pytesSetCellLabel(String& label, uint16_t value) {
```
Comment (AndreasBoehm) [inline src/PytesCanReceiver.cpp:10]: we are reading `uint8` from the can message but here you defined `uint16`?

File: src/PytesCanReceiver.cpp [hunk 8]
```diff
@@ -284,6 +307,175 @@ void PytesCanReceiver::onMessage(twai_message_t rx_message)
...
+                        _stats->_cellMinVoltageName.c_str(), _stats->_cellMaxVoltageName.c_str());
+            }
+            break;
+        }
+
+        case 0x402: { // Pytes protocol: Highest/Lowest Cell Temperature
+            _stats->setVoltage(this->scaleValue(this->readSignedInt16(rx_message.data), 0.01), millis());
+            _stats->_cellMaxTemperature = this->scaleValue(this->readUnsignedInt16(rx_message.data), 0.1);
+            _stats->_cellMinTemperature = this->scaleValue(this->readUnsignedInt16(rx_message.data + 2), 0.1);
+            pytesSetCellLabel(_stats->_cellMaxTemperatureName, this->readUnsignedInt8(rx_message.data + 4));
```
Comment (AndreasBoehm) [inline src/PytesCanReceiver.cpp:328]: pytes protocol reports the name as an int and victron as a string?
I am wondering why we need to use different methods to get the name as a string.

File: src/PytesCanReceiver.cpp [hunk 9]
```diff
@@ -284,6 +307,175 @@ void PytesCanReceiver::onMessage(twai_message_t rx_message)
...
+            if (_verboseLogging) {
+                MessageOutput.printf("[Pytes] internalFailure: %d (bits: %08x)\r\n",
+                        _stats->_alarmInternalFailure, alarmBits);
+            }
+            break;
+        }
+
+        case 0x408: { // Pytes protocol: charge status (similar to Pylontech msg 0x35e)
+            bool chargeEnabled = rx_message.data[0];
+            bool dischargeEnabled = rx_message.data[1];
```
Comment (AndreasBoehm) [inline src/PytesCanReceiver.cpp:431]: We should use this information and not throw it away, if we need it, because with the victron protcol we don't  get the values like that, but we get a count of modules that are blocking charge or blocking discharge. Does something similar exist for the message `0x40b`? Right now we only parse module count online and offline.

I know that the comment is not directly related to this line of code but also: https://github.com/helgeerbe/OpenDTU-OnBattery/pull/1196/files#diff-fdc211e7100fe5cfbe1c29df40fefa664212330184f5ec0394748fe4637808ceR455

Comment (AndreasBoehm) [reply src/PytesCanReceiver.cpp:431]: Right now we show 0 for both `Charging blocked` and `Discharging blocked` under `Battery modules` even though we don't populate this field when the pytes protocol is used

**Review by AndreasBoehm — CHANGES_REQUESTED** [review]

**Comment by AndreasBoehm** [comment]
I switched to the PYTES protocol and the web ui is showing a strange value for `voltage` sometimes, the value shown was `1,70 V`. In the live view on top and in the battery card on the bottom.

It happens randomly and i could not find anything strange in the CAN messages, they all looked fine and showed the correct voltage of 53,xx

Did you have the same experience @ranma ?

File: src/PytesCanReceiver.cpp (hunk 7 above)
Comment (ranma) [reply src/PytesCanReceiver.cpp:10]: Oops, that should have been readUnsignedInt16() where this is called.
One byte is the cell id, the second byte is presumably the module id (though I only have a single module, so I can't easily verify that currently).

File: src/PytesCanReceiver.cpp (hunk 8 above)
Comment (ranma) [reply src/PytesCanReceiver.cpp:328]: Correct. In victron protocol the battery converts to string internally, e.g.:
```
  case 9:
                    /* msg 0x374 */
    bVar8 = *(byte *)&batteryState->field35_min_cell_mv_id;
    *pbVar7 = bVar8 / 10 + 0x30;
    pbVar7[1] = bVar8 % 10 + 0x30;
    uVar9 = (uint)batteryState->field35_min_cell_mv_id >> 8 & 0xff;
    cVar1 = (char)(uVar9 / 10);
    pbVar7[2] = cVar1 + 0x30;
    pbVar7[3] = (char)uVar9 + cVar1 * -10 + 0x30;
    pbVar7[4] = 0;
    pbVar7[5] = 0;
    pbVar7[6] = 0;
    pbVar7[7] = 0;
    return 8;
```

**Comment by ranma** [comment]
> I switched to the PYTES protocol and the web ui is showing a strange value for `voltage` sometimes, the value shown was `1,70 V`. In the live view on top and in the battery card on the bottom.

I don't remember seeing any drops like that, not in the UI nor in the MQTT data.

File: src/PytesCanReceiver.cpp (hunk 9 above)
Comment (ranma) [reply src/PytesCanReceiver.cpp:431]: Ah, yes. I though I had updated the code already at some point, maybe I accidentally reverted that.

FWIW I've repeatedly updated this comment with the protocol details: https://github.com/helgeerbe/OpenDTU-OnBattery/issues/1188#issuecomment-2307087491

> Does something similar exist for the message 0x40b

Its part of message 0x408, which has a bitfield which indicates which modules have charge inhibited (byte 5) / discharge inhibited (byte 6).

**Comment by ranma** [comment]
> I switched to the PYTES protocol and the web ui is showing a strange value for `voltage` sometimes, the value shown was `1,70 V`. In the live view on top and in the battery card on the bottom.

I can currently reproduce this in testing right now, I sometimes see `2.90 V` instead of `53.15 V`.

**Comment by ranma** [comment]
> I can currently reproduce this in testing right now, I sometimes see `2.90 V` instead of `53.15 V`.

There was a stray setVoltage() fatfingered into the message 0x402 handler at some point, removed that to fix the issue.

Also verified the charge inhibited count with replayed messages from when the battery was full, it shows `Charging blocked     1` now:
image

File: src/PytesCanReceiver.cpp [hunk 10]
```diff
@@ -248,7 +278,7 @@ void PytesCanReceiver::onMessage(twai_message_t rx_message)
             _stats->_totalCapacity = this->readUnsignedInt16(rx_message.data);
```
Comment (AndreasBoehm) [inline src/PytesCanReceiver.cpp:278]: I think this is only available in the Victron protocol, right?

Comment (ranma) [reply src/PytesCanReceiver.cpp:278]: Its there for Pytes protocol:
```
        case 0x409: { // Pytes protocol: full mAh / remaining mAh
            _stats->_totalCapacity = this->scaleValue(this->readUnsignedInt32(rx_message.data), 0.001);
```
